File: trs-sdk-typescript/tests/client.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import http from "node:http";
import net from "node:net";

import { TRSClient, TRSConnectionError, TRSValidationError } from "../dist/index.js";

function startServer(delayMs = 0) {
  const records = new Map();
  const server = http.createServer(async (req, res) => {
    if (delayMs) {
      await new Promise((resolve) => setTimeout(resolve, delayMs));
    }
    const chunks = [];
    for await (const chunk of req) chunks.push(chunk);
    const bodyText = Buffer.concat(chunks).toString("utf8") || "{}";
    const body = JSON.parse(bodyText);

    const send = (code, payload) => {
      const text = JSON.stringify(payload);
      res.writeHead(code, { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(text) });
      res.end(text);
    };

    if (req.method === "GET" && req.url === "/health") return send(200, { status: "ok", runtime: "1.0.0", node: "0.1.0" });
    if (req.method === "POST" && req.url === "/submit") {
      const record = body.record;
      if (!record || typeof record !== "object") return send(422, { detail: "field 'record' must be an object" });
      if (!record.payload || typeof record.payload !== "object" || !("subject" in record.payload || "goal" in record.payload || "action" in record.payload)) {
        return send(200, { accepted: false, record_id: record.id ?? "", errors: ["5.3 Payload Shape"] });
      }
      records.set(record.id, record);
      return send(200, { accepted: true, record_id: record.id ?? "", errors: [] });
    }
    if (req.method === "POST" && req.url === "/query") {
      const q = body.query ?? {};
      let out = Array.from(records.values());
      if (q.type) out = out.filter((r) => r.type === q.type);
      return send(200, { records: out });
    }
    if (req.method === "POST" && req.url === "/sync") {
      const incoming = Array.isArray(body.records) ? body.records : [];
      incoming.forEach((r) => records.set(r.id, r));
      return send(200, { accepted_count: incoming.length, rejected_count: 0, appended_ids: incoming.map((r) => r.id), rejected_errors: [] });
    }
    if (req.method === "POST" && req.url === "/replay") return send(200, { coordination: { unresolved_intentions: [] } });
    return send(404, { error: "not found" });
  });
  return new Promise((resolve) => {
    server.listen(0, "127.0.0.1", () => {
      const address = server.address();
      resolve({ server, url: `http://127.0.0.1:${address.port}` });
    });
  });
}

test("health, submit, query, sync, replay", async () => {
  const { server, url } = await startServer();
  try {
    const client = new TRSClient(url);
    const health = await client.health();
    assert.equal(health.status, "ok");
    await client.submit({ id: "g1", type: "Observation", payload: { subject: "boot", value: 1 } });
    const intentions = await client.query({ type: "Observation" });
    assert.equal(intentions.length, 1);
    const sync = await client.sync(intentions);
    assert.equal(sync.accepted_count, 1);
    const replay = await client.replay();
    assert.ok(replay.coordination);
  } finally {
    server.close();
  }
});

test("submit invalid raises TRSValidationError", async () => {
  const { server, url } = await startServer();
  try {
    const client = new TRSClient(url);
    await assert.rejects(async () => client.submit({ id: "bad", type: "Observation", payload: {} }), TRSValidationError);
  } finally {
    server.close();
  }
});

test("timeout raises TRSConnectionError", async () => {
  const socket = net.createServer();
  await new Promise((resolve) => socket.listen(0, "127.0.0.1", resolve));
  const address = socket.address();
  const port = address.port;
  await new Promise((resolve) => socket.close(resolve));
  const client = new TRSClient(`http://127.0.0.1:${port}`, 20);
  await assert.rejects(async () => client.health(), TRSConnectionError);
});

File: trs-sdk-go/trs/client_test.go
package trs

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func testServer(delay time.Duration) *httptest.Server {
	records := map[string]map[string]any{}
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if delay > 0 {
			time.Sleep(delay)
		}
		writeJSON := func(code int, payload any) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(code)
			_ = json.NewEncoder(w).Encode(payload)
		}

		switch {
		case r.Method == http.MethodGet && r.URL.Path == "/health":
			writeJSON(200, map[string]any{"status": "ok", "runtime": "1.0.0", "node": "0.1.0"})
		case r.Method == http.MethodPost && r.URL.Path == "/submit":
			var body map[string]any
			_ = json.NewDecoder(r.Body).Decode(&body)
			record, _ := body["record"].(map[string]any)
			payload, _ := record["payload"].(map[string]any)
			if _, hasSubject := payload["subject"]; !hasSubject {
				if _, hasGoal := payload["goal"]; !hasGoal {
					if _, hasAction := payload["action"]; !hasAction {
						writeJSON(200, map[string]any{"accepted": false, "record_id": record["id"], "errors": []string{"5.3 Payload Shape"}})
						return
					}
				}
			}
			id, _ := record["id"].(string)
			records[id] = record
			writeJSON(200, map[string]any{"accepted": true, "record_id": id, "errors": []string{}})
		case r.Method == http.MethodPost && r.URL.Path == "/query":
			writeJSON(200, map[string]any{"records": mapsToAnySlice(records)})
		case r.Method == http.MethodPost && r.URL.Path == "/sync":
			var body map[string]any
			_ = json.NewDecoder(r.Body).Decode(&body)
			incoming, _ := body["records"].([]any)
			writeJSON(200, map[string]any{
				"accepted_count":  len(incoming),
				"rejected_count":  0,
				"appended_ids":    []string{"g1"},
				"rejected_errors": [][]string{},
			})
		case r.Method == http.MethodPost && r.URL.Path == "/replay":
			writeJSON(200, map[string]any{"coordination": map[string]any{"unresolved_intentions": []string{}}})
		default:
			writeJSON(404, map[string]any{"error": "not found"})
		}
	}))
}

func mapsToAnySlice(values map[string]map[string]any) []any {
	out := make([]any, 0, len(values))
	for _, value := range values {
		out = append(out, value)
	}
	return out
}

func TestClientFlow(t *testing.T) {
	s := testServer(0)
	defer s.Close()

	client := NewClient(s.URL, 5*time.Second)
	ctx := context.Background()

	health, err := client.Health(ctx)
	if err != nil {
		t.Fatalf("health failed: %v", err)
	}
	if health.Status != "ok" {
		t.Fatalf("unexpected health status: %s", health.Status)
	}

	_, err = client.Submit(ctx, map[string]any{
		"id":      "g1",
		"type":    "Observation",
		"payload": map[string]any{"subject": "boot", "value": 1},
	})
	if err != nil {
		t.Fatalf("submit failed: %v", err)
	}

	records, err := client.Query(ctx, map[string]any{})
	if err != nil {
		t.Fatalf("query failed: %v", err)
	}
	if len(records) != 1 {
		t.Fatalf("expected 1 record, got %d", len(records))
	}

	sync, err := client.Sync(ctx, records)
	if err != nil {
		t.Fatalf("sync failed: %v", err)
	}
	if sync.AcceptedCount != 1 {
		t.Fatalf("unexpected accepted count: %d", sync.AcceptedCount)
	}

	replay, err := client.Replay(ctx)
	if err != nil {
		t.Fatalf("replay failed: %v", err)
	}
	if _, ok := replay["coordination"]; !ok {
		t.Fatalf("missing coordination in replay payload")
	}
}

func TestSubmitValidationError(t *testing.T) {
	s := testServer(0)
	defer s.Close()

	client := NewClient(s.URL, 5*time.Second)
	_, err := client.Submit(context.Background(), map[string]any{
		"id":      "bad",
		"type":    "Observation",
		"payload": map[string]any{},
	})
	if err == nil || !IsValidationError(err) {
		t.Fatalf("expected validation error, got %v", err)
	}
}

func TestTimeoutConnectionError(t *testing.T) {
	s := testServer(200 * time.Millisecond)
	defer s.Close()

	client := NewClient(s.URL, 20*time.Millisecond)
	_, err := client.Health(context.Background())
	if err == nil || !IsConnectionError(err) {
		t.Fatalf("expected connection error, got %v", err)
	}
}
